GET /platform implementation

louis-ver opened the pull request:
- Implementation of `GET /platform` endpoint
 - `ErrorCodeAndMessage` schema
 - `PlatformProperties` schema
 - Setting of PlatformProperties through `server-config.json`, along with its validation

louis-ver committed:
Add platform endpoint and config validation

louis-ver committed:
Add testing of properties validation

louis-ver committed:
Add validation for https support

louis-ver commented:
I looked into the reason why the 2.7 build is not passing.

The problem is with "type hints" in functions. For example `ErrorCodeAndMessage`, the constructor is:
```
def __init__(self, error_code: int = 0, error_message: str = None):
    # init code here
```

The equivalent notation for 2.7 would be (1)

```
def __init__(self, error_code=0, error_message=None):
    # init code here

__init__.__annotations__ = {'error_code': int, 'error_message': str}
```

Or (2)

```
def __init__(self, error_code=0, error_message=None):
    # type: (int, str)
    # init code here
```

I suggest we use the 2.7 compatible notation in order to not lose 2.7 support. As a result, all methods/functions will have to use this notation.

simon-dube commented:
@glatard What are your thoughts on Python 2.7 support for type hints? Without this syntax, the project won't build in Python 2.7. Otherwise, we can leave out the type hints all together, but I think this would be a significant drawback not to have type hints.

glatard commented:
Good question... I would rather go with type hints and drop Python 2.7, since this is a new project and Python 2.7 won't be supported from 2020. But I might miss important arguments, I created a Twitter poll to get more insights. Do you see any good reason to keep Python 2.7?

louis-ver committed:
Remove unnecessary “is True” check to assert

louis-ver committed:
Remove 2.7 build in CI

louis-ver commented:
Based on a conversation @glatard and I just had, I think it's safe to drop support for Python 2.7 from this PR moving forward. Commits will no longer be tested against Python 2.7, but rather Python 3.5 and later.

Since the CARMIN server is a new project, and is aimed mostly towards Server Administrators and advanced users, there seems to be no good reason to support Python 2.7.

louis-ver committed:
Remove testing for pyp 5.3.1

simon-dube commented on carmin-server/server/common/exceptions.py:
```diff
@@ -0,0 +1,3 @@
+class MissingRequiredParameterError(Exception):
```
I would name this MissingRequiredParameterException instead of Error.

louis-ver replied:
[According to PEP8](https://www.python.org/dev/peps/pep-0008/#exception-names), this is the proper way of naming Exceptions.

simon-dube commented on carmin-server/server/properties_validation.py:
```diff
@@ -0,0 +1,52 @@
+def properties_validation(config_data=None):
+    """Performs validation on server_config.json file. Checks for required
+    parameters and supported options."""
+    supported_protocols = [
```
I would specify these information in a private config file instead of in the function on its own.

simon-dube commented on carmin-server/server/properties_validation.py:
```diff
@@ -0,0 +1,52 @@
+    supported_protocols = [
+        "http", "https", "ftp", "sftp", "ftps", "scp", "webdav"
+    ]
+    supported_modules = [
```
I would specify these information in a private config file instead of in the function on its own.

simon-dube commented on carmin-server/server/test/test_server_config.py:
```diff
@@ -0,0 +1,86 @@
+from server import properties_validation
+from server.common.exceptions import MissingRequiredParameterError
+
+CONFIG_DATA = {
```
Could isolate a bunch of test config_data in their own module instead of directly in the test. Could be reused in the case of system tests requiring a running server for example.

simon-dube reviewed (changes requested):
Looks good. Just a few things to improve upon to have an even better structure. GJ.

louis-ver committed:
Moved enums into config.py

Also, converted config_data in tests so that it is now a fixture, and is reusable throughout the test suite

louis-ver commented:
Changes made. Thank you for the review 👍
- `supported_modules` and `supported_protocols` now in `config.py`
- `CONFIG_DATA` now a fixture that can be used across the `test` module

louis-ver committed:
Fix readme

louis-ver committed:
Call superclass constructor

simon-dube commented on carmin-server/server/properties_validation.py:
```diff
@@ -3,24 +3,20 @@
 import os
 import json
+from typing import Dict
+from .config import SUPPORTED_PROTOCOLS
```
Can be imported in a single line.

simon-dube reviewed (approved):
Good for me. Just maybe bring the two imports in a single line. Good job!

louis-ver committed:
Import into single line
